### ISSUE TITLE
Getting Backend failed to start up, RangeError [ERR_SOCKET_BAD_PORT]: Port should be >= 0 and < 65536. Received NaN. while deploying with k8s

### ISSUE THREAD

**Issue by suraj-zemoso** [backstage/backstage#5633]
I'm trying to deploy the backstage with k8s but getting following error

`Backend failed to start up, RangeError [ERR_SOCKET_BAD_PORT]: Port should be >= 0 and < 65536. Received NaN.`

## Expected Behavior
Deployment should work as expected and pod should be in running state

## Current Behavior
Deployment is failing and pod is restarting due to the mentioned error.

## Steps to Reproduce

1. Dockerize the backstage app
2. Create deployment with `kubectl` 
3. Check deployment status and pod status with `kubectl -n backstage get po, deploy`

## Context
1. app-config.yml
```
app:
  title: App
  baseUrl: http://localhost:7000

organization:
  name: Org

backend:
  baseUrl: http://localhost:7000
  listen:
    port: 7000
```

2. k8s deployment config
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: backstage
  namespace: backstage
spec:
  replicas: 1
  selector:
    matchLabels:
      app: backstage
  template:
    metadata:
      labels:
        app: backstage
    spec:
      containers:
        - name: backstage
          image: private-reg/org/backstage
          imagePullPolicy: Always
          ports:
            - containerPort: 7000
              name: http
          envFrom:
            - configMapRef:
                name: postgres-config
            - configMapRef:
                name: backstage-cm
            - secretRef:
                name: backstage-secrets
      imagePullSecrets:
      - name: regcred
```

After deployment, pod is restarting with logs as follows

`2021-05-11T05:49:18.868Z backstage info Loaded config from app-config.yaml 
2021-05-11T05:49:18.872Z backstage info Created UrlReader predicateMux{readers=azure{host=dev.azure.com,authed=false},bitbucket{host=bitbucket.org,authed=false},github{host=github.com,authed=true},gitlab{host=gitlab.com,authed=false},fetch{} 
Backend failed to start up, RangeError [ERR_SOCKET_BAD_PORT]: Port should be >= 0 and < 65536. Received NaN.`
## Your Environment

- NodeJS Version (v14):
- Operating System and Version (e.g. Ubuntu 20.04):
- Browser Information:

**Comment by jhaals** [comment]
Hi!

Sorry to hear that you're having problems starting the container.
It looks like backstage is unable to find/parse the port section of your config. I tried to reproduce this locally and get errors if I set a too high port numer or true/false as port but I have managed to get `Received NaN` so something is probably wrong with the configuration file. I'd try overwriting the entrypoint in your kubernetes manifest to instead run a long `sleep` so that you can verify the contents of the app-config.yaml inside the container.

**Comment by adamdmharvey** [comment]
I'll note your front end AND backend in your sample `app-config.yaml` are defined as running on the same port.  Generally, one is 3000 (frontend/UI) and one is 7000 (backend) out of the box - might this be the reason why one is failing to start? (port in use?)

**Comment by Crevil** [comment]
What are the contents of the `backstage-cm` ConfigMap?

**Comment by suraj-zemoso** [comment]
> What are the contents of the `backstage-cm` ConfigMap?

Using it to pass some extra config values of auth0 and okta.

Here are the contents
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: backstage-cm
  namespace: backstage
data:
  AUTH_AUTH0_DOMAIN: org.auth0.com
  AUTH_OKTA_DOMAIN: https://org.okta.com
```

**Comment by Crevil** [comment]
Ok. Can you start the dockerimage in a container locally with eg. `docker run`?

As you haven't specificed any runtime injected configuration, I would expect it to fail there as well, and this not be related to the kubernetes part.

**Comment by suraj-zemoso** [comment]
> Ok. Can you start the dockerimage in a container locally with eg. `docker run`?
> 
> As you haven't specificed any runtime injected configuration, I would expect it to fail there as well, and this not be related to the kubernetes part.

It's actually working fine when  running in a container

```
$ docker run --name backstage --env-file=./env.list --link postgres backstage:latest 
2021-05-15T11:12:42.967Z backstage info Loaded config from app-config.yaml 
2021-05-15T11:12:42.973Z backstage info Created UrlReader predicateMux{readers=azure{host=dev.azure.com,authed=false},bitbucket{host=bitbucket.org,authed=false},github{host=github.com,authed=true},gitlab{host=gitlab.com,authed=false},fetch{} 
2021-05-15T11:12:43.539Z catalog info Locations Refresh: Beginning locations refresh type=plugin component=catalog-all-locations-refresh
2021-05-15T11:12:43.547Z catalog info Locations Refresh: Visiting 1 locations type=plugin component=catalog-all-locations-refresh
2021-05-15T11:12:43.548Z catalog info Locations Refresh: Refreshing location bootstrap:bootstrap type=plugin component=catalog-all-locations-refresh
2021-05-15T11:12:43.549Z catalog info Reading GitHub users and groups type=plugin
2021-05-15T11:12:44.206Z auth info Configuring provider, github type=plugin
2021-05-15T11:12:44.208Z auth info Configuring provider, auth0 type=plugin
2021-05-15T11:12:44.209Z auth info Configuring provider, okta type=plugin
2021-05-15T11:12:44.211Z techdocs info Creating Local publisher for TechDocs type=plugin
2021-05-15T11:12:44.214Z proxy info [HPM] Proxy created: [Function: filter]  ->  https://example.com type=plugin
2021-05-15T11:12:44.215Z proxy info [HPM] Proxy rewrite rule created: "^/api/proxy/test/" ~> "/" type=plugin
2021-05-15T11:12:44.217Z proxy info [HPM] Proxy created: [Function: filter]  ->  https://sonarcloud.io/api type=plugin
2021-05-15T11:12:44.217Z proxy info [HPM] Proxy rewrite rule created: "^/api/proxy/sonarqube/" ~> "/" type=plugin
2021-05-15T11:12:44.219Z app info Serving static app content from /app/packages/app/dist type=plugin
2021-05-15T11:12:44.385Z app info Injecting env config into module-backstage.47ee3fe0.js type=plugin
2021-05-15T11:12:44.391Z backstage info Initializing http server 
2021-05-15T11:12:44.393Z backstage info Listening on :7000 
2021-05-15T11:13:29.538Z catalog info Read 55 entities from location bootstrap:bootstrap in 46.0s type=plugin component=catalog-all-locations-refresh
2021-05-15T11:13:29.650Z catalog info Wrote 55 entities from location bootstrap:bootstrap in 112ms type=plugin component=catalog-all-locations-refresh
2021-05-15T11:13:29.652Z catalog info Locations Refresh: Completed locations refresh in 46.1s type=plugin component=catalog-all-locations-refresh
```

**Comment by Crevil** [comment]
What is in `env.list` in this run?

**Comment by suraj-zemoso** [comment]
> What is in `env.list` in this run?

it contains some environment variables like the one that I've mentioned in configMap

**Comment by Crevil** [comment]
Cool. Then it sure seems related to kubernetes. Thanks for clearing that up 👍

Kubernetes will set a whole lot of environment variables in a pod for service discovery purposes. Do you have any kubernetes Services in the same namespace? I would guess at least a one for backstage it self?

I guess backstage could read one of those variables somehow?

As your `app-config.yaml` does not reference anything from the env, it seems like that is unintended.

**Comment by suraj-zemoso** [comment]
> Cool. Then it sure seems related to kubernetes. Thanks for clearing that up 
> 
> Kubernetes will set a whole lot of environment variables in a pod for service discovery purposes. Do you have any kubernetes Services in the same namespace? I would guess at least a one for backstage it self?
> 
> I guess backstage could read one of those variables somehow?
> 
> As your `app-config.yaml` does not reference anything from the env, it seems like that is unintended.

Yes. I've config file for backstage service. As I'm testing this deployment on my local cluster, I've used a NodePort
```
apiVersion: v1
kind: Service
metadata:
  name: backstage
  namespace: backstage
spec:
  selector:
    app: backstage
  type: NodePort
  ports:
    - name: backstage-port
      port: 7000
      targetPort: 7000
      nodePort: 30002
```

**Comment by Crevil** [comment]
What happens if you set a nodePort below 65536? Just to test.

**Comment by suraj-zemoso** [comment]
> What happens if you set a nodePort below 65536? Just to test.

Actually the issue is with the deployment. Service is getting created without any issue.

Here is the steps that I'm following,
1. Creating the ConfigMap (and secrets as well)
2. Creating the deployment (where the pod is failing and getting the error mentioned in the ticket) - (backstage-deployment.yml)
3. Creating the backstage service (backstage-service.yml)

**Comment by Crevil** [comment]
The reason I'm curious is, that services results in environment variables in deployments :)

So adding a service like `backstage` with port 12345 would set an env variable in the pod like `BACKSTAGE_PORT=12345` or something like that. I'm not completely sure on the naming scheme.

But just to be sure I understand correctly, the pod fails right upon deployment creation, ie. before the service is created?

**Comment by suraj-zemoso** [comment]
> The reason I'm curious is, that services results in environment variables in deployments :)
> 
> So adding a service like `backstage` with port 12345 would set an env variable in the pod like `BACKSTAGE_PORT=12345` or something like that. I'm not completely sure on the naming scheme.
> 
> But just to be sure I understand correctly, the pod fails right upon deployment creation, ie. before the service is created?

yeah. It's failing upon the deployment creation.

**Comment by Crevil** [comment]
I've done some digging. The error originates from Node.js's `net` module, more specifically the [`listen` method](https://github.com/nodejs/node/blob/cb2bdc632a802eed7ee8c5c9388811de73029602/lib/net.js#L1438) of the `Server` class. We call this method from backend-common's [`ServiceBuilderImpl`](https://github.com/backstage/backstage/blob/master/packages/backend-common/src/service/lib/ServiceBuilderImpl.ts#L182).

From what I can see in the code the `listen(port, host, ...)` call gets its `port` argument from the following places.

- `DEFAULT_PORT=7000`: https://github.com/backstage/backstage/blob/master/packages/backend-common/src/service/lib/ServiceBuilderImpl.ts#L43
- `this.port`: This can be set in two ways
  - Call to method `setPort()`
  - Call to method `loadConfig` which reads `app-config.yaml` files entries
    - `backend.listen.port`
    - `backend.listen` if it's a `string` value where it is parsed as `host:port`

Judging from your provided `app-config.yaml` we would expect `port` to be `7000` like the default value.

This does not really bring me that much closer to the cause I think I just wanted to share my findings.

**Comment by Crevil** [comment]
Actually I think I'm wrong in where the error originates. The error is `RangeError [ERR_SOCKET_BAD_PORT]: Port should be >= 0 and < 65536. Received NaN.`. The [`validatePort` function](https://github.com/nodejs/node/blob/cb2bdc632a802eed7ee8c5c9388811de73029602/lib/internal/validators.js#L202) has a `name` argument which defaults to `Port`, which is seen in the error as well. This means `validatePort` is called without a second argument.
The only reference I can find in `Node.js` is the [`lookupAndConnect` function](https://github.com/nodejs/node/blob/cb2bdc632a802eed7ee8c5c9388811de73029602/lib/net.js#L992) which is used by the [`Socket#connect` method](https://github.com/nodejs/node/blob/cb2bdc632a802eed7ee8c5c9388811de73029602/lib/net.js#L968).

I can see that `backend-common` creates an HTTP server which logs `backstage info Initializing http server`, but this is not present in your provided log snippets, so this indicates that it is in fact not the HTTP server that causes the failure.

Maybe it's the catalog? In the logs from your docker image we can see the line just after startup `catalog info Locations Refresh` which is missing in the k8s case. 🤔

**Comment by Crevil** [comment]
The docker image run you provided looks like it has both github auth and some catalog bootstrap locations. Is the provided `app-config.yaml` a redacted version maybe? If such, can you provide us with the full version with any secret details omitted ofc.

**Comment by suraj-zemoso** [comment]
> The docker image run you provided looks like it has both github auth and some catalog bootstrap locations. Is the provided `app-config.yaml` a redacted version maybe? If such, can you provide us with the full version with any secret details omitted ofc.

Sure. Here is the full version of `app-config.yml`
```
app:
  title: App
  baseUrl: http://localhost:7000

organization:
  name: org

backend:
  baseUrl: http://localhost:7000
  listen:
    port: 7000
  csp:
    connect-src: ["'self'", 'http:', 'https:']
  cors:
    origin: http://localhost:7000
    methods: [GET, POST, PUT, DELETE]
    credentials: true
  # config options: https://node-postgres.com/api/client
  database:
    client: pg
    connection:
      host:
        $env: POSTGRES_HOST
      port:
        $env: POSTGRES_PORT
      user:
        $env: POSTGRES_USER
      password:
        $env: POSTGRES_PASSWORD
      # https://node-postgres.com/features/ssl
      #ssl: require # see https://www.postgresql.org/docs/current/libpq-ssl.html Table 33.1. SSL Mode Descriptions (e.g. require)
        #ca: # if you have a CA file and want to verify it you can uncomment this section
        #  $file: <file-path>/ca/server.crt
  # workingDirectory: /tmp # Use this to configure a working directory for the scaffolder, defaults to the OS temp-dir

integrations:
  github:
    - host: github.com
      token:
        $env: GITHUB_TOKEN
    ### Example for how to add your GitHub Enterprise instance using the API:
    # - host: ghe.example.net
    #   apiBaseUrl: https://ghe.example.net/api/v3
    #   token:
    #     $env: GHE_TOKEN

proxy:
  '/test':
    target: 'https://example.com'
    changeOrigin: true
  '/sonarqube':
    target: https://sonarcloud.io/api
    allowedMethods: [ 'GET' ]
    headers:
      Authorization:
        # Environmental variable: SONARQUBE_AUTH_HEADER
        # Value: 'Basic base64("<sonar-auth-token>:")'
        # Encode the "<sonar-auth-token>:" string using base64 encoder.
        # Note the trailing colon (:) at the end of token.
        # Example of the value: Basic bXktYXBpLWtleTo=
        # Fetch the sonar-auth-token from https://sonarcloud.io/account/security/
        $env: SONARQUBE_AUTH_HEADER

# Reference documentation http://backstage.io/docs/features/techdocs/configuration
# Note: After experimenting with basic setup, use CI/CD to generate docs
# and an external cloud storage when deploying TechDocs for production use-case.
# https://backstage.io/docs/features/techdocs/how-to-guides#how-to-migrate-from-techdocs-basic-to-recommended-deployment-approach
techdocs:
  builder: 'local' # Alternatives - 'external'
  generators:
    techdocs: 'docker' # Alternatives - 'local'
  publisher:
    type: 'local' # Alternatives - 'googleGcs' or 'awsS3'. Read documentation for using alternatives.

auth:
  # see https://backstage.io/docs/tutorials/quickstart-app-auth to know more about enabling auth providers
  providers:
    github:
      development:
        clientId:
          $env: AUTH_GITHUB_CLIENT_ID
        clientSecret:
          $env: AUTH_GITHUB_CLIENT_SECRET
        ## uncomment the following two lines if using enterprise
        # enterpriseInstanceUrl:
        #  $env: AUTH_GITHUB_ENTERPRISE_INSTANCE_URL
    auth0:
      development:
        clientId:
          $env: AUTH_AUTH0_CLIENT_ID
        clientSecret:
          $env: AUTH_AUTH0_CLIENT_SECRET
        domain:
          $env: AUTH_AUTH0_DOMAIN
    okta:
      development:
        clientId:
          $env: AUTH_OKTA_CLIENT_ID
        clientSecret:
          $env: AUTH_OKTA_CLIENT_SECRET
        audience:
          $env: AUTH_OKTA_DOMAIN

scaffolder:
  github:
    token:
      $env: GITHUB_TOKEN
    visibility: internal # or 'public' or 'private'

catalog:
  rules:
    - allow: [Component, API, Group, User, Template, Location]
  locations:
    # organization
    - type: github-org
      target: https://github.com/org

    # Org Backstage Templates
    - type: url
      target: https://github.com/org/backstage-templates/blob/main/react-ts-storybook/template.yaml
      rules:
        - allow: [Template]

    - type: url
      target: https://github.com/org/backstage-templates/blob/main/spring-starter-todo/template.yaml
      rules:
        - allow: [Template]

    # Backstage example components
    - type: url
      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-components.yaml

    # Backstage example APIs
    - type: url
      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-apis.yaml

    # Backstage example organization groups
    - type: url
      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/acme/org.yaml
      rules:
        - allow: [Group, User]

    # Backstage example templates
    - type: url
      target: https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/sample-templates/react-ssr-template/template.yaml
      rules:
        - allow: [Template]
    - type: url
      target: https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/sample-templates/springboot-grpc-template/template.yaml
      rules:
        - allow: [Template]
    - type: url
      target: https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/sample-templates/create-react-app/template.yaml
      rules:
        - allow: [Template]
    - type: url
      target: https://github.com/spotify/cookiecutter-golang/blob/master/template.yaml
      rules:
        - allow: [Template]
    - type: url
      target: https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/sample-templates/docs-template/template.yaml
      rules:
        - allow: [Template]
  processors:
    githubOrg:
        providers:
          - target: https://github.com
            apiBaseUrl: https://api.github.com
            token:
              $env: GITHUB_TOKEN

```

**Comment by suraj-zemoso** [comment]
Any update for me ?
Did you got the root cause of the issue?

**Comment by Rugvip** [comment]
@suraj-zemoso I'd say it's tricky to continue without some more information. You probably want to troubleshoot it a bit yourself using the debugger or some logging, making sure that you see the values you expect and try to pinpoint where there error is coming from.

**Comment by suraj-zemoso** [comment]
We finally found the issue.

It's because of port conflicts.

We tried using the different port and the issue got resolved.

Thanks you so much for providing the valuable feedback on this issue and thanks for all suggestions which helped me to locate the bug and resolution.

I'm closing the bug now as it's no longer related to backstage configuration.

Thank you all

**Comment by kalinon** [comment]
@suraj-zemoso what do you mean? Im encountering the same issue... which i cannot figure out.

**Comment by dariawhitecastle** [comment]
We experienced this issue and the root cause seems to be managing resources in the same namespace across unique releases. We fixed this by clearing the secret for POSTGRES_PORT env var in the namespace that was incorrectly overridden by a typo in one of the releases. Whether or not you use k8s secrets for your deployments, double-check that environment variables are getting set properly in the container. 

`docker exec container printenv POSTGRES_PORT`

or for k8s

`kubectl get secret your-secret-name -o jsonpath='{.data.POSTGRES_PORT}' | base64 --decode`
